File: src/api/routers/streaming.py
"""
Streaming API endpoints.

Provides generic streaming endpoints for all agents and transports.
"""

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from typing import Dict, Any

from src.models.streaming.session import CreateSessionRequest, SessionResponse
from src.models.streaming.message import MessageRequest, MessageResponse
from src.services.streaming.streaming_service import StreamingService

try:
    from src.providers.logger import get_logger
    logger = get_logger(__name__)
except ImportError:
    import logging
    logger = logging.getLogger(__name__)

router = APIRouter()

# Global streaming service instance
_streaming_service: StreamingService = None


def get_streaming_service() -> StreamingService:
    """Get or create streaming service instance."""
    global _streaming_service
    if _streaming_service is None:
        _streaming_service = StreamingService()
    return _streaming_service


@router.get("/agents")
async def get_available_agents():
    """
    Get list of available streaming agents.
    
    Returns all agents that can be used for streaming sessions.
    """
    try:
        streaming_service = get_streaming_service()
        agents = streaming_service.list_available_agents()
        
        return {
            "agents": agents,
            "count": len(agents)
        }
        
    except Exception as e:
        logger.error(f"Error getting available agents: {e}")
        raise HTTPException(status_code=500, detail="Failed to get available agents")


@router.post("/sessions", response_model=SessionResponse)
async def create_session(request: CreateSessionRequest):
    """
    Create a new streaming session.
    
    Creates a session for the specified agent with the chosen transport.
    """
    try:
        logger.info(f"Creating session for agent {request.agent_id}")
        
        streaming_service = get_streaming_service()
        session_id = await streaming_service.create_session(
            agent_id=request.agent_id,
            transport_type=request.transport_type,
            user_context=request.user_context
        )
        
        # Get session info for response
        session_info = streaming_service.get_session_info(session_id)
        if not session_info:
            raise HTTPException(status_code=500, detail="Session created but info not available")
        
        session_data = session_info["session"]
        
        return SessionResponse(
            session_id=session_data["session_id"],
            agent_id=session_data["agent_id"],
            agent_name=session_data["agent_name"],
            status=session_data["status"],
            created_at=session_data["created_at"],
            transport_type=session_data["transport_type"]
        )
        
    except ValueError as e:
        logger.warning(f"Invalid session creation request: {e}")
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error creating session: {e}")
        raise HTTPException(status_code=500, detail="Failed to create session")


@router.get("/sessions/{session_id}/events")
async def stream_events(session_id: str):
    """
    Stream events for a session via Server-Sent Events.
    
    Establishes an SSE connection to receive real-time events from the agent.
    """
    try:
        logger.info(f"Creating event stream for session {session_id}")
        
        streaming_service = get_streaming_service()
        
        # Validate session exists
        session_info = streaming_service.get_session_info(session_id)
        if not session_info:
            raise HTTPException(status_code=404, detail=f"Session {session_id} not found")
        
        async def event_generator():
            try:
                async for event in streaming_service.get_event_stream(session_id):
                    yield event
            except Exception as e:
                logger.error(f"Error in event stream for session {session_id}: {e}")
                # Send error event and close stream
                error_event = f"data: {{\"event_type\": \"error\", \"data\": {{\"message\": \"Stream error: {str(e)}\"}}}}\n\n"
                yield error_event
        
        return StreamingResponse(
            event_generator(),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "X-Accel-Buffering": "no"  # Disable nginx buffering
            }
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error setting up event stream for session {session_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to create event stream")


@router.post("/sessions/{session_id}/messages", response_model=MessageResponse)
async def send_message(session_id: str, request: MessageRequest):
    """
    Send a message to an agent in a streaming session.
    
    The agent will process the message and stream response events.
    """
    try:
        logger.info(f"Sending message to session {session_id}: {request.message[:100]}...")
        
        streaming_service = get_streaming_service()
        
        # Validate session exists and is active
        session_info = streaming_service.get_session_info(session_id)
        if not session_info:
            raise HTTPException(status_code=404, detail=f"Session {session_id} not found")
        
        if not session_info["is_active"]:
            raise HTTPException(status_code=410, detail=f"Session {session_id} is not active")
        
        # Process message (this will stream events through the transport)
        await streaming_service.process_message(session_id, request.message)
        
        return MessageResponse(
            status="sent",
            message_id=f"msg_{session_id}_{int(time.time() * 1000)}"
        )
        
    except HTTPException:
        raise
    except ValueError as e:
        logger.warning(f"Invalid message request for session {session_id}: {e}")
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error processing message for session {session_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to process message")


@router.get("/sessions/{session_id}")
async def get_session_info(session_id: str):
    """
    Get information about a specific session.
    
    Returns session details, agent info, and transport status.
    """
    try:
        streaming_service = get_streaming_service()
        session_info = streaming_service.get_session_info(session_id)
        
        if not session_info:
            raise HTTPException(status_code=404, detail=f"Session {session_id} not found")
        
        return session_info
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting session info for {session_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get session info")


@router.delete("/sessions/{session_id}")
async def close_session(session_id: str):
    """
    Close a streaming session and clean up resources.
    
    Terminates the session and releases associated resources.
    """
    try:
        logger.info(f"Closing session {session_id}")
        
        streaming_service = get_streaming_service()
        success = await streaming_service.close_session(session_id)
        
        if not success:
            raise HTTPException(status_code=404, detail=f"Session {session_id} not found")
        
        return {"status": "closed", "session_id": session_id}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error closing session {session_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to close session")


@router.get("/stats")
async def get_streaming_stats():
    """
    Get streaming service statistics.
    
    Returns information about active sessions, agents, and performance metrics.
    """
    try:
        streaming_service = get_streaming_service()
        stats = streaming_service.get_service_stats()
        
        return stats
        
    except Exception as e:
        logger.error(f"Error getting streaming stats: {e}")
        raise HTTPException(status_code=500, detail="Failed to get streaming stats")


# Import time for message IDs
import time

File: ui/src/lib/streaming.ts
/* eslint-disable @typescript-eslint/no-explicit-any -- TODO: Fix types gradually */
/**
 * Streaming client for knowledge agents.
 * 
 * Provides a generic streaming client that handles SSE connections,
 * message sending, and session management for knowledge agents.
 */

import { getApiBaseUrl } from './environment';
import { getAuthHeader } from './auth';

export interface StreamingEvent {
  event_type: string;
  data: any;
  timestamp: string;
  turn_complete?: boolean;
  session_id?: string;
}

export interface SessionInfo {
  session_id: string;
  agent_id: string;
  agent_name: string;
  status: string;
  created_at: string;
  transport_type: string;
}

export interface AgentInfo {
  id: string;
  name: string;
  description: string;
  capabilities: string[];
}

export class StreamingClient {
  private eventSource: EventSource | null = null;
  private sessionId: string | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 5000; // 5 seconds
  private onEventCallback: ((event: StreamingEvent) => void) | null = null;
  private onConnectionCallback: ((connected: boolean) => void) | null = null;

  /**
   * Create a new streaming session for a knowledge agent.
   */
  async createSession(agentId: string): Promise<string> {
    try {
      // Get auth header for API request
      const authHeader = await getAuthHeader();
      
      const response = await fetch(`${getApiBaseUrl()}/api/v1/knowledge-agents/chat/sessions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...authHeader,
        },
        body: JSON.stringify({
          agent_id: agentId,
        }),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({ detail: 'Unknown error' }));
        throw new Error(`Failed to create session: ${errorData.detail || response.statusText}`);
      }

      const data = await response.json();
      this.sessionId = data.session_id;
      
      if (!this.sessionId) {
        throw new Error('Session ID not received from server');
      }
      
      console.log('Created streaming session:', data);
      return this.sessionId;
    } catch (error) {
      console.error('Error creating session:', error);
      throw error;
    }
  }

  /**
   * Connect to the SSE event stream.
   */
  connectStream(
    onEvent: (event: StreamingEvent) => void,
    onConnection?: (connected: boolean) => void
  ): void {
    if (!this.sessionId) {
      throw new Error('No active session. Call createSession() first.');
    }

    this.onEventCallback = onEvent;
    this.onConnectionCallback = onConnection || null;
    
    // Start the async SSE connection
    this._connectSSE().catch(error => {
      console.error('Failed to connect SSE:', error);
      this.onConnectionCallback?.(false);
    });
  }

  /**
   * Send a message to the knowledge agent.
   * Automatically creates a session if none exists.
   */
  async sendMessage(message: string, agentId?: string): Promise<void> {
    // Auto-create session if needed
    if (!this.sessionId && agentId) {
      console.log('No active session, creating one automatically...');
      await this.createSession(agentId);
    }
    
    if (!this.sessionId) {
      throw new Error('No active session and no agentId provided to create one.');
    }

    try {
      // Get auth header for API request
      const authHeader = await getAuthHeader();
      
      const response = await fetch(
        `${getApiBaseUrl()}/api/v1/knowledge-agents/chat/sessions/${this.sessionId}/messages`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            ...authHeader,
          },
          body: JSON.stringify({
            message,
          }),
        }
      );

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({ detail: 'Unknown error' }));
        throw new Error(`Failed to send message: ${errorData.detail || response.statusText}`);
      }

      console.log('Message sent successfully');
    } catch (error) {
      console.error('Error sending message:', error);
      throw error;
    }
  }

  /**
   * Close the streaming connection and session.
   */
  async close(): Promise<void> {
    // Close SSE connection
    if (this.eventSource) {
      this.eventSource.close();
      this.eventSource = null;
    }

    // Close session on server
    if (this.sessionId) {
      try {
        const authHeader = await getAuthHeader();
        await fetch(
          `${getApiBaseUrl()}/api/v1/knowledge-agents/chat/sessions/${this.sessionId}`,
          {
            method: 'DELETE',
            headers: {
              ...authHeader
            }
          }
        );
        console.log('Session closed successfully');
      } catch (error) {
        console.error('Error closing session:', error);
      }
      
      this.sessionId = null;
    }

    // Reset state
    this.reconnectAttempts = 0;
    this.onEventCallback = null;
    this.onConnectionCallback = null;
  }

  /**
   * Get information about the current session.
   */
  async getSessionInfo(): Promise<SessionInfo | null> {
    if (!this.sessionId) {
      return null;
    }

    try {
      // Get auth header for API request
      const authHeader = await getAuthHeader();
      
      const response = await fetch(
        `${getApiBaseUrl()}/api/v1/knowledge-agents/chat/sessions/${this.sessionId}`,
        {
          headers: authHeader,
        }
      );

      if (!response.ok) {
        return null;
      }

      const data = await response.json();
      return data.session;
    } catch (error) {
      console.error('Error getting session info:', error);
      return null;
    }
  }

  /**
   * Get list of available knowledge agents.
   */
  static async getAvailableAgents(): Promise<AgentInfo[]> {
    try {
      // Get auth header for API request
      const authHeader = await getAuthHeader();
      
      const response = await fetch(`${getApiBaseUrl()}/api/v1/knowledge-agents/chat/agents`, {
        headers: authHeader,
      });
      
      if (!response.ok) {
        throw new Error(`Failed to get agents: ${response.statusText}`);
      }

      const data = await response.json();
      return data.agents;
    } catch (error) {
      console.error('Error getting available agents:', error);
      throw error;
    }
  }

  /**
   * Check if client is connected.
   */
  isConnected(): boolean {
    return this.eventSource !== null && this.eventSource.readyState === EventSource.OPEN;
  }

  /**
   * Get current session ID.
   */
  getSessionId(): string | null {
    return this.sessionId;
  }

  private async _connectSSE(): Promise<void> {
    if (!this.sessionId) {
      console.error('Cannot connect SSE without session ID');
      return;
    }

    try {
      console.log(`Connecting to SSE stream for session: ${this.sessionId}`);
      
      // Get auth header for SSE request
      const authHeader = await getAuthHeader();
      
      // Use fetch instead of EventSource to support authentication headers
      const response = await fetch(
        `${getApiBaseUrl()}/api/v1/knowledge-agents/chat/sessions/${this.sessionId}/events`,
        {
          headers: {
            'Accept': 'text/event-stream',
            'Cache-Control': 'no-cache',
            ...authHeader,
          },
        }
      );

      if (!response.ok) {
        throw new Error(`SSE connection failed: ${response.status} ${response.statusText}`);
      }

      if (!response.body) {
        throw new Error('Response body is null');
      }

      // Create a manual SSE reader
      const reader = response.body.getReader();
      const decoder = new TextDecoder();

      // Create mock EventSource for compatibility
      this.eventSource = {
        readyState: 1, // OPEN
        close: () => {
          reader.cancel();
          this.eventSource = null;
        }
      } as EventSource;

      // Signal connection opened
      console.log('SSE connection opened');
      this.reconnectAttempts = 0;
      this.onConnectionCallback?.(true);

      // Process SSE stream manually
      this._processSSEStream(reader, decoder);

    } catch (error) {
      console.error('Error setting up SSE connection:', error);
      this.onConnectionCallback?.(false);
      
      // Auto-reconnect logic
      if (this.reconnectAttempts < this.maxReconnectAttempts) {
        this.reconnectAttempts++;
        console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts})...`);
        
        setTimeout(() => {
          this._connectSSE();
        }, this.reconnectDelay);
      }
    }
  }

  private async _processSSEStream(reader: ReadableStreamDefaultReader<Uint8Array>, decoder: TextDecoder): Promise<void> {
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        
        if (done) {
          console.log('SSE stream ended');
          this.onConnectionCallback?.(false);
          break;
        }

        // Decode the chunk and add to buffer
        buffer += decoder.decode(value, { stream: true });

        // Process complete SSE events
        const lines = buffer.split('\n');
        buffer = lines.pop() || ''; // Keep incomplete line in buffer

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const data = line.substring(6); // Remove 'data: '
              if (data.trim()) {
                const streamingEvent: StreamingEvent = JSON.parse(data);
                console.log('Received SSE event:', streamingEvent);
                this.onEventCallback?.(streamingEvent);
              }
            } catch (error) {
              console.error('Error parsing SSE event:', error);
            }
          }
        }
      }
    } catch (error) {
      console.error('Error reading SSE stream:', error);
      this.onConnectionCallback?.(false);
      
      // Auto-reconnect logic
      if (this.reconnectAttempts < this.maxReconnectAttempts) {
        this.reconnectAttempts++;
        console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts})...`);
        
        setTimeout(() => {
          this._connectSSE();
        }, this.reconnectDelay);
      }
    } finally {
      reader.releaseLock();
    }
  }
}
